pycalphad/io/cs_dat.py

```diff
@@ -1159,6 +1162,999 @@ def parse_cs_dat(instring):
+            flat_constituents = [constituent for sublattice in constituents for constituent in sublattice]
+
+            # Get constituent mapping
+            constituent_mapping = make_constituent_mapping(constituents, endmember_params)
+
+            # Get sublattice info
+            sublattices = dbf.phases[phase_name].sublattices
+            nSublattices = len(sublattices)
+
+            # Write sublattice information
+            if phase_model in ('SUBL','SUBLM'):
+                # Number of sublattices and weights only for SUBL
+                output += f'{nSublattices:4}\n'
+                output += f'  {"      ".join([f"{weight:.5f}" for weight in sublattices])}\n'
+            output += f'{"".join([f"{len(sub):4}" for sub in constituents])}\n'
+
+            # Write constituent names
+            for sub in constituents:
+                output += f'  {"".join([f"{constituent.capitalize():25}" for constituent in sub])}\n'
+
+            # Write charge magnitudes and chemical groups for MQM
+            if phase_model in ('SUBG', 'SUBQ'):
+                for ion in ['cations','anions']:
+                    charges = [abs(species.charge) for species in chemical_groups[ion].keys()]
+                    groups = [chemical_groups[ion][species] for species in chemical_groups[ion].keys()]
+                    # Write charges for ion type
+                    output += f'  {"      ".join([f"{charges[i]:.5f}" for i in range(len(charges))])}\n'
+                    # Write chemical groups for ion type
+                    output += f'   {"".join([f"{groups[i]:4}" for i in range(len(charges))])}\n'
+
+            # Write constituent-to-endmember pairing arrays
+            for sub in constituent_mapping:
+                output += f'{"".join([f"{constituent:4}" for constituent in sub])}\n'
+
+            # Write quadruplet coordination numbers for MQM
+            if phase_model in ('SUBG', 'SUBQ'):
+                detect_query = (
+                    (where("phase_name") == phase_name) & \
+                    (where("parameter_type") == "MQMZ")
+                )
+                params = dbf._parameters.search(detect_query)
+                for param in params:
+                    con = []
+                    for ion in param['constituent_array']:
+                        for species in ion:
+                            con.append(flat_constituents.index(species.name) + 1)
+                    con_string = ''.join([f"{c:4}" for c in con])
+                    z_string = '      '.join([f"{z:.7f}" for z in param['coordinations']])
+                    output += f'{con_string}  {z_string}\n'
+
+        # Write magnetic excess mixing data
+        if phase_model in ('RKMPM', 'SUBLM'):
+            # Get excess magnetic parameters
+            for tc in tcs:
+                tc_value = tc['parameter']
+                tc_constituents = tc['constituent_array']
+                tcs.remove(tc)
```

Comment:
Mutating a list while you're iterating over it can have unexpected results in Python. I think you could sort both lists of parameters by the `constituent_array` key. Something like (may need more error checking)
```python
from collections import defaultdict
magnetic_by_constituent_array = defaultdict(list)
for tc in tcs:
    magnetic_by_constituent_array[tc['constituent_array']].append(tc)
for bmagn in bmagns:
    magnetic_by_constituent_array[bmagn['constituent_array']].append(bmagn)

for tc, bmagn in magnetic_by_constituent_array.values():
    print(tc, bmagn) # etc
```